### PR TITLE
Cleaner init and shutdown

### PR DESCRIPTION
**Pull request by Pierre-Felber**
initGLFW was called twice, after shutdown the driver could not be reinstantiated.
Some use cases need complete shutdown of the glDriver, and bring it up only once in a while, saving memory and cpu cycles.

### Description:
Allow complete shutdown and later reinstantiation.

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [ ] Public APIs match existing style.
- [ ] Any breaking changes have a deprecation path or have been discussed.

### CODE REVIEW

**Comment by andydotxyz** [comment]
Thanks for this. Is the use-case that you call app.Quit(), but then call Run() again later on?
Perhaps you could write a test to show this working, so we can avoid accidentally breaking it again in the future?

**Comment by Pierre-Felber** [comment]
> Thanks for this. Is the use-case that you call app.Quit(), but then call Run() again later on?
Exactly
> Perhaps you could write a test to show this working, so we can avoid accidentally breaking it again in the future?
Will do

**Comment by Pierre-Felber** [comment]
> Will do
Done

**Comment by andydotxyz** [comment]
Test 1 is more likely useful, as test 2 is less close to testing the intent.
However it loads applications in a goroutine which is not supported.
Is there another way to test this?

**Comment by Pierre-Felber** [comment]
> However it loads applications in a goroutine which is not supported.

I think there is a missunderstanding, regarding 'Main-Thread'. As far as I understand, all calls to the GL library from init to shutdown must happen from within the **SAME** OS-Thread (because of the libraries usage of thread local storage), which is accomplished by calling runtime.LockOSThread. 

So it should not matter at all, if the GoRoutine that is calling into GL library is the first one created (the one that you seem to regard the Main-GoRoutine), or any other. I wrote the test especially this way to underline my understanding, that it is of no importance whether the GoRoutine that is initializing the GL library is the first, second, or millionth created one. 

Besides that, of course the test could be  implemented, so that the GoRoutine that is created first, is the one calling into the GL library, though I actually think it is better, to clarify the actual requirements than just copy'n paste the sentence 'all calls must come from the MainThread' which I find obscuring. What is the MainThread anyway? That's totally up to the programmer, to define that.

For me as the consumer of the fyne package it is important to know, what are the requirements, and when I red 'MainThread' I got confused, since to me, there are no threads in GO just goroutines, that get scheduled to run on anyone thread of an underlying pool of os-threads. And, if I need to, I can create an os thread that is reserved to run just one specific go routine by calling runtime.LockOSThread. 

Maybe I am missing something? The test passes fyne here, and the application that I am writing is running fyne.App instances in goroutines just fyne...

**Comment by andydotxyz** [comment]
Unfortunately what you describe is not quite the case. the App.Run() or Window.ShowAndRun() must be called from main() (or a routine called directly from it - not a goroutine). Some operating systems are looser on this requirement so if you are on Linux it may work, but on others, like macOS, the application will not function.

"Main Thread" is really synonymous with the main thread of execution - yes it is not literally a thread but we were trying to differentiate from being within a spawned goroutine.

**Comment by Pierre-Felber** [comment]
> Unfortunately what you describe is not quite the case. the App.Run() or Window.ShowAndRun() must be called from main() (or a routine called directly from it - not a goroutine). Some operating systems are looser on this requirement so if you are on Linux it may work, but on others, **like macOS**, the application will not function.

Thanks for the explanation. I will look at this.

**Comment by Pierre-Felber** [comment]
> As discussed in thread, looking for more realistic test of the re-use

Please advise me. To me it looks like I'd need to create a specific test package for the test case of reuse, since all tests are run in a their own go routine. To test reuse, I need to implement the test case in TestMain, which is only allowed once per package. Could you tell me where I should place the test package and how it should be named.

**Comment by andydotxyz** [comment]
Perhaps you can remind us of the code that does not work that this fixes? We can base a test on that so we know it's testing the right thing.

File: internal/driver/glfw/loop.go [hunk 1]
```diff
@@ -65,13 +65,17 @@ func (d *gLDriver) runGL() {
...
-	d.initGLFW()
+	initOnce.Do(d.initGLFW)
 
 	for {
 		select {
 		case <-d.done:
+			runMutex.Lock()
 			fps.Stop()
 			glfw.Terminate()
+			runFlag = false
```
Comment (Pierre-Felber) [inline internal/driver/glfw/loop.go]: was left true

File: internal/driver/glfw/loop.go [hunk 2]
```diff
@@ -65,13 +65,17 @@ func (d *gLDriver) runGL() {
...
+			initOnce = &sync.Once{}
```
Comment (Pierre-Felber) [inline internal/driver/glfw/loop.go]: not creating a new Instance would result in not initializing the glfw lib again, obviously...

**Comment by andydotxyz** [comment]
> To test reuse, I need to implement the test case in TestMain, which is only allowed once per package. Could you tell me where I should place the test package and how it should be named.

The testing has moved on a bit since this and now I think it will be possible.
In the tests for `glfw` package there is a driver instance `d` (set up in window_test.go).
This is running on the main thread as you say.
I suspect it would be possible to do something here - for example emit a signal after d.Run() then call d.Run() again. 
Your test case would call d.Quit() and then assert that the signal you expect was seen.
Depending on the nature of the failure to launch the second instance it may require another check to show the replacement loop is running correctly.

Sorry it's been so long, and sorry that is not a complete solution, but I hope it helps.
It would be a good idea to get this into 1.3 if possible as others have noted the issue more recently.

**Comment by Jacalz** [comment]
I think that small part of this PR was included into the v1.3 release. Would you like to still work on this @Pierre-Felber, if so perhaps this could still be updated against the conflicts and the comment from Andy above? :)

**Comment by andydotxyz** [comment]
Many of the issues underlying this have been fixed and this PR has been inactive for a while.
It would be easier at this point to implement the remaining changes as a new PR.

**Pull request closed without merging**